cunumeric/module.py

```diff
@@ -2498,6 +2499,139 @@ def take(
+    ----------
+    arr : ndarray (Ni..., M, Nk...)
+        Source array
+    indices : ndarray (Ni..., J, Nk...)
+        Indices to take along each 1d slice of `arr`. This must match the
+        dimension of arr, but dimensions Ni and Nj only need to broadcast
+        against `arr`.
+    axis : int
+        The axis to take 1d slices along. If axis is None, the input array is
+        treated as if it had first been flattened to 1d, for consistency with
+        :func:`cunumeric.sort` and :func:`cunumeric.argsort`.
+
+    Returns
+    -------
+    out: ndarray (Ni..., J, Nk...)
+        The indexed result.
+
+    See Also
+    --------
+    numpy.take_along_axis
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise TypeError("`indices` must be an integer array")
+
+    if axis is None:
+        if indices.ndim != 1:
+            raise ValueError("indices must be 1D if axis=None")
+        a = a.ravel()
+        axis = 0
+    else:
+        axis = normalize_axis_index(axis, a.ndim)
+
+    ndim = a.ndim
+    if ndim != indices.ndim:
+        raise ValueError(
+            "`indices` and `a` must have the same number of dimensions"
+        )
+    return a[_fill_fancy_index_for_along_axis_routines(a.shape, axis, indices)]
+
+
+@add_boilerplate("a", "indices", "values")
+def put_along_axis(
+    a: ndarray, indices: ndarray, values: ndarray, axis: int
+) -> ndarray:
+    """
+    Put values into the destination array by matching 1d index and data slices.
+
+    This iterates over matching 1d slices oriented along the specified axis in
+    the index and data arrays, and uses the former to place values into the
+    latter. These slices can be different lengths.
+
+    Functions returning an index along an axis, like :func:`cunumeric.argsort`
+    and :func:`cunumeric.argpartition`, produce suitable indices for
+    this function.
+
+    Parameters
+    ----------
+    arr : ndarray (Ni..., M, Nk...)
+        Destination array.
+    indices : ndarray (Ni..., J, Nk...)
+        Indices to change along each 1d slice of `arr`. This must match the
+        dimension of arr, but dimensions in Ni and Nj may be 1 to broadcast
+        against `arr`.
+    values : array_like (Ni..., J, Nk...)
+        values to insert at those indices. Its shape and dimension are
+        broadcast to match that of `indices`.
+    axis : int
+        The axis to take 1d slices along. If axis is None, the destination
+        array is treated as if a flattened 1d view had been created of it.
+
+    See Also
+    --------
+    numpy.put_along_axis
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
+
+    if axis is None:
+        if indices.ndim != 1:
+            raise ValueError("indices must be 1D if axis=None")
+        a = a.ravel()
```

Comment:
This can return a copy, meaning the user-supplied `a` will end up not being modified. We should check here that `a.ravel()` indeed created a view. I created a PR adding such a method: https://github.com/nv-legate/cunumeric/pull/474.

Alternatively, we can require that if `axis` is missing then `a` must be 1d (and update the documentation accordingly). Any preference @magnatelee ?

For what it's worth, it looks like the current implementation will create a copy even if reshape just drops a singleton dimension (if run with `-cunumeric:test`, so we go through `DeferredArray`):

```
a = cn.ones((12,1))
b = a.ravel()
print(a.same_root(b))  # prints `False`
```

Reply:
Numpy documentation doesn't say it has to return the view, but I believe it does actually return the view because it uses `a.flat`
Cupy, however, calls a.ravel().
I think we can still support the case when axis=None and a.ndim>0 returning a copy  as far as we document this. Supporting some case partially can be better than not supporting it at all.

Reply:
> Numpy documentation doesn't say it has to return the view

On the NumPy documentation I see: "If axis is None, the destination array is treated as if a flattened 1d *view* had been created of it." So it looks like NumPy guarantees a view even if `axis` is `None`.

> I think we can still support the case when axis=None and a.ndim>1 returning a copy

The function has no return value, its only effect is updating the passed array.

Reply:
Yes, you are right, I didn't check the lines and though this comment is for `take_along_axis`. I can disable this use case until we get  `a.flat` implemented

Reply:
Yes, let's error out for now when `axis is None` and `a.ndim > 1`. But please leave a TODO note, to revisit this once #474 or something equivalent is merged.